File: frontend-angular/src/app/core/domain/use-cases/apply-status-effect.use-case.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { BattlePort } from '../../ports/battle.port';
import { Battle } from '../models/battle.model';

@Injectable({
  providedIn: 'root',
})
export class ApplyStatusEffectUseCase {
  constructor(private readonly battlePort: BattlePort) {}

  execute(battleId: string, creatureId: string, effect: string, action: 'ADD' | 'REMOVE'): Observable<Battle> {
    return this.battlePort.applyStatusEffect(battleId, creatureId, effect, action);
  }
}

File: backend/src/main/kotlin/de/thomcz/pap/battle/backend/domain/port/in/ApplyStatusEffectUseCase.kt
package de.thomcz.pap.battle.backend.domain.port.`in`

import de.thomcz.pap.battle.backend.application.dto.ApplyStatusEffectCommand
import de.thomcz.pap.battle.backend.domain.model.Battle
import java.util.UUID

/**
 * Use case: Add or remove a status effect on a creature in an active battle.
 *
 * Input: Battle ID, creature ID, effect name, action (ADD/REMOVE)
 * Output: Updated Battle aggregate with StatusEffectApplied or StatusEffectRemoved event
 */
interface ApplyStatusEffectUseCase {
    fun execute(battleId: UUID, creatureId: UUID, command: ApplyStatusEffectCommand, userId: String): Battle
}
